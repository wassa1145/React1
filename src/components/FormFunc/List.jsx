import React from 'react';
import { Message } from './Message';

export const List = (props) => {
  return (
    <div className="list" ref={props.refEl}>
      <ul>
        {props.messages.map((message, index) => (
          <li
            key={index}
            className={message.systemMessage ? 'system-message' : 'message'}
          >
            <Message message={message} />
          </li>
        ))}
      </ul>
    </div>
  );
};
