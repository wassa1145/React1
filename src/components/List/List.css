.list {
	width: 100%;
  padding: 20px;
  margin: 0 auto;
  box-sizing: border-box;
  flex-grow: 1;
  overflow-y: auto;
}
ul {
  display: flex !important;
  flex-direction: column;
}
ul,p {
  list-style-type: none;
  padding: 0;
  margin: 0;
}
.message,
.system-message {
  text-align: center;
  border-radius: 17px;
  background-color: #fff;
  padding: 10px;
  font-size: 18px;
  width: 80%;
  position: relative;
  margin-bottom: 20px;
  box-sizing: border-box;
  box-shadow: 0 0 10px rgb(0 0 0 / 20%);
}
.main-dark .message,
.main-dark .system-message {
  color: #fff !important;
  background-color: rgb(153, 150, 150) !important;
}
.message {
  align-self: flex-end;
}
.system-message {
  align-self: flex-start;
}
.message__author,
.message__text {
  text-align: start;
}
.message__author {
  font-weight: bold;
}