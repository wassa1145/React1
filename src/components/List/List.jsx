import React from 'react';
import './List.css';

export const List = (props) => (
  <div className="list" ref={props.refEl} data-testid="messages-list">
    <ul>
      {props.messages.map((message, index) => (
        <li
          key={index}
          className={message.systemMessage ? 'system-message' : 'message'}
          role="message"
        >
          <p className="message__author">{message.author}</p>
          <p className="message__text">{message.message}</p>
        </li>
      ))}
    </ul>
  </div>
);
