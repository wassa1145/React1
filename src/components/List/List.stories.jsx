import React from 'react';

import { List } from './List';

export default {
  title: 'MyComponents/List',
  component: List,
  // argTypes: {
  //   onListClick: { action: 'click' },
  // },
};

const Template = (args) => <List {...args} />;

export const Messages = Template.bind({});
Messages.args = {
  messages: [
    { author: 'User', message: 'Сообщение 1' },
    {
      author: 'Mr. Robot',
      message: 'Мы получили ваше сообщение! В ближайшее время мы вам ответим.',
      systemMessage: true,
    },
    { author: 'User', message: 'Сообщение 2' },
    {
      author: 'Mr. Robot',
      message: 'Мы получили ваше сообщение! В ближайшее время мы вам ответим.',
      systemMessage: true,
    },
  ],
};

export const Empty = Template.bind({});
Empty.args = {
  messages: [],
};
