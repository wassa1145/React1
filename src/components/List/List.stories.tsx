import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { List } from './List';

export default {
  title: 'MyComponents/List',
  component: List,
  // argTypes: {
  //   onListClick: { action: 'click' },
  // },
} as ComponentMeta<typeof List>;

const Template: ComponentStory<typeof List> = (args) => <List {...args} />;

export const Messages = Template.bind({});
Messages.args = {
  messages: [
    { author: 'User', value: 'Сообщение 1', id: '1' },
    {
      author: 'Mr. Robot',
      value: 'Мы получили ваше сообщение! В ближайшее время мы вам ответим.',
      systemMessage: true,
      id: '2',
    },
    { author: 'User', value: 'Сообщение 2', id: '3' },
    {
      author: 'Mr. Robot',
      value: 'Мы получили ваше сообщение! В ближайшее время мы вам ответим.',
      systemMessage: true,
      id: '4',
    },
  ],
};

export const Empty = Template.bind({});
Empty.args = {
  messages: [],
};
