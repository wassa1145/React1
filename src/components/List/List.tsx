import React, { FC } from 'react';
import './List.css';

interface Message {
  id: string;
  author: string;
  message: string;
  systemMessage?: boolean;
}

interface MessagesListProps {
  messages: Message[];
  listEl?: React.RefObject<HTMLInputElement>;
}

export const List: FC<MessagesListProps> = ({ messages, listEl }) => (
  <div className="list" data-testid="messages-list" ref={listEl}>
    <ul>
      {messages.map((message) => (
        <li
          key={message.id}
          className={message.systemMessage ? 'system-message' : 'message'}
          role="message"
        >
          <p className="message__author">{message.author}</p>
          <p className="message__text">{message.message}</p>
        </li>
      ))}
    </ul>
  </div>
);
