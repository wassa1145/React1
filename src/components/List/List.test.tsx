import React from 'react';
import { List } from './List';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('List', () => {
  it('render component', () => {
    const messages = [
      { author: 'Bot', value: 'new message', id: '1' },
      { author: 'User', value: 'user message', id: '2' },
    ];
    render(<List messages={messages} />);
    expect(screen.getByTestId('messages-list')).toBeInTheDocument();
  });

  it('render with snapshot', () => {
    const messages = [
      { author: 'Bot', value: 'new message', id: '1' },
      { author: 'User', value: 'user message', id: '2' },
    ];
    const { asFragment } = render(<List messages={messages} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('check rendering message', () => {
    const messages = [{ author: 'Bot', value: 'test', id: '1' }];
    render(<List messages={messages} />);
    expect(screen.getByText(/test/)).toBeInTheDocument();
    expect(screen.getByText(/Bot/)).toBeInTheDocument();
  });

  it('count messages', () => {
    const messages = [
      { author: 'Bot', value: 'new message', id: '1' },
      { author: 'User', value: 'new message2', id: '2' },
      { author: 'User', value: 'new message3', id: '3' },
    ];
    render(<List messages={messages} />);
    expect(screen.queryAllByRole('message').length).toBe(3);
  });
});
