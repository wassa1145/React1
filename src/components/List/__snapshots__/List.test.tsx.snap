// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List render with snapshot 1`] = `
<DocumentFragment>
  <div
    class="list"
    data-testid="messages-list"
  >
    <ul>
      <li
        class="message"
        role="message"
      >
        <p
          class="message__author"
        >
          Bot
        </p>
        <p
          class="message__text"
        >
          new message
        </p>
      </li>
      <li
        class="message"
        role="message"
      >
        <p
          class="message__author"
        >
          User
        </p>
        <p
          class="message__text"
        >
          user message
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
